File: pinyougou-web/pinyougou-user-web/src/main/java/com/pinyougou/user/controller/UserController.java
package com.pinyougou.user.controller;

import com.alibaba.dubbo.config.annotation.Reference;
import com.pinyougou.pojo.User;
import com.pinyougou.service.UserService;
import org.apache.commons.codec.digest.DigestUtils;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import java.util.HashMap;
import java.util.Map;

/**
 * 用户控制器
 *
 * @author lee.siu.wah
 * @version 1.0
 * <p>File Created at 2019-03-16<p>
 */
@RestController
@RequestMapping("/user")
public class UserController {

    @Reference(timeout = 10000)
    private UserService userService;

    /**
     * 用户注册
     */
    @PostMapping("/save")
    public boolean save(@RequestBody User user, String code) {
        try {
            // 检验短信验证码
            boolean flag = userService.checkSmsCode(user.getPhone(), code);
            if (flag) {
                userService.save(user);
            }
            return flag;
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return false;
    }

    /**
     * 更改密码
     */
    @PostMapping("/updatePassword")
    public boolean updatePassword(@RequestBody Map<String, Object> content) {
        try {
            String oldPassword = (String) content.get("oldPassword");
            String md5OldPassword = DigestUtils.md5Hex(oldPassword);
            String password = (String) content.get("password");
            String loginName = this.getLoginName();
            String okPassword = userService.findByUsername("password", loginName);
            if (md5OldPassword.equals(okPassword)) {
                try {
                    User user = new User();
                    // 密码需要MD5加密 commons-codec-xxx.jar
                    String md5Password = DigestUtils.md5Hex(password);
                    user.setPassword(md5Password);
                    user.setUsername(loginName);

                    userService.update("password", user, md5Password);
                } catch (Exception ex) {
                    ex.printStackTrace();
                    return false;
                }
            } else {
                return false;
            }
            return true;
        } catch (Exception ex) {
            ex.printStackTrace();
            return false;
        }
    }

    /**
     * 获取原手机号
     */
    @GetMapping("/findPhoneNumber")
    public Map<String, Object> findPhoneNumber() {
        String loginName = this.getLoginName();
        String phone = userService.findByUsername("phone", loginName);
        String num1 = phone.substring(0, 3);
        String num2 = phone.substring(7);
        String showNumber = num1 + "****" + num2;
        Map<String, Object> map = new HashMap<>();
        map.put("phone", phone);
        map.put("showNumber", showNumber);
        return map;
    }

    /**
     * 检验身份
     */
    @PostMapping("/checkPhoneNumber")
    public boolean checkPhoneNumber(@RequestBody User user, String code, String phoneCode, HttpServletRequest request) {
        // 判断请求方式
        if ("post".equalsIgnoreCase(request.getMethod())) {
            // 判断验证码
            String oldCode = (String) request.getSession().getAttribute(VerifyController.VERIFY_CODE);
            if (code != null && code.equalsIgnoreCase(oldCode)) {
                try {
                    // 检验短信验证码
                    return userService.checkSmsCode(user.getPhone(), phoneCode);
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        }
        return false;
    }

    /**
     * 发送短信验证码
     */
    @GetMapping("/sendSmsCode")
    public boolean sendSmsCode(String phone) {
        try {
            return userService.sendSmsCode(phone);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return false;
    }

    /**
     * 更新手机号
     */
    @PostMapping("/updatePhoneNumber")
    public boolean updatePhoneNumber(@RequestBody User user, String code, String phoneCode, HttpServletRequest request) {
        // 判断请求方式
        if ("post".equalsIgnoreCase(request.getMethod())) {
            // 判断验证码
            String oldCode = (String) request.getSession().getAttribute(VerifyController.VERIFY_CODE);
            if (code != null && code.equalsIgnoreCase(oldCode)) {
                try {
                    // 检验短信验证码
                    boolean flag = userService.checkSmsCode(user.getPhone(), phoneCode);
                    if (flag) {
                        String loginName = this.getLoginName();
                        user.setUsername(loginName);
                        userService.update("phone", user, user.getPhone());
                    }
                    return flag;
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        }
        return false;
    }

    public String getLoginName() {
        SecurityContext context = SecurityContextHolder.getContext();

        return context.getAuthentication().getName();
    }
}

File: pinyougou-web/pinyougou-user-web/src/main/webapp/js/controller/userController.js
/** 定义控制器层 */
app.controller('userController', function ($scope, $timeout, baseService) {

    // 定义json对象
    $scope.user = {}

    // 用户注册
    $scope.save = function () {
        // 判断密码是否一致
        if ($scope.okPassword && $scope.user.password == $scope.okPassword) {
            // 发送异步请求
            baseService.sendPost("/user/save?code=" + $scope.code,
                $scope.user).then(function (response) {
                // 获取响应数据
                if (response.data) {
                    // 跳转到登录页面
                    // 清空表单数据
                    $scope.user = {};
                    $scope.okPassword = "";
                    $scope.code = "";
                } else {
                    alert("注册失败！");
                }
            });
        } else {
            alert("两次密码不一致！");
        }
    };
    /** 修改密码 */
    $scope.updatePassword = function () {
        // 判断密码是否一致
        if (!$scope.content.password || $scope.content.password != $scope.content.confirm_password) {
            alert("两次密码不一致！")
            return;
        }
        if ($scope.content.password == $scope.content.oldPassword) {
            alert("新密码不能与原密码一致！")
            return;
        }
        // 发送异步请求
        baseService.sendPost("/user/updatePassword", $scope.content).then(function (response) {
            // 获取响应数据
            if (response.data) {
                // 跳转到登录页面
                alert("密码修改成功!")
                location.href = "http://sso.pinyougou.com/login?service=http%3A%2F%2Fuser.pinyougou.com%2Flogin"
                // 清空表单数据
                $scope.content.oldPassword = "";
                $scope.content.password = "";
                $scope.content.confirm_password = "";
            } else {
                alert("密码更新失败！");
            }
        });
    }
    /** 获取手机号码 */
    $scope.getPhoneNumber = function () {
        // 发送异步请求
        baseService.sendGet("/user/findPhoneNumber").then(function (response) {
            // 获取响应数据
            $scope.user.phone = response.data.phone;
            $scope.showNumber = response.data.showNumber
        });
    }
    /** 验证身份 */
    $scope.checkPhoneNumber = function () {
        // 发送异步请求
        baseService.sendPost("/user/checkPhoneNumber?code=" + $scope.code + "&phoneCode=" + $scope.phoneCode,
            $scope.user).then(function (response) {
            // 获取响应数据
            if (response.data) {
                // 跳转到第二步 绑定新手机号
                location.href = "/home-setting-address-phone.html";
                // 清空表单数据
                $scope.phoneCode = "";
                $scope.code = "";
            } else {
                alert("验证失败！");
            }
        });
    }
    $scope.updatePhoneNumber =function () {
        // 发送异步请求
        baseService.sendPost("/user/updatePhoneNumber?code=" + $scope.code + "&phoneCode=" + $scope.phoneCode,
            $scope.user).then(function (response) {
            // 获取响应数据
            if (response.data) {
                // 跳转到第二步 绑定新手机号
                location.href = "home-setting-address-complete.html";
                // 清空表单数据
                $scope.user = "";
                $scope.phoneCode = "";
                $scope.code = "";
            } else {
                alert("绑定失败！");
            }
        });
    }
    // 定义显示文本
    $scope.tipMsg = "获取短信验证码";
    $scope.flag = false;

    // 发送短信验证码
    $scope.sendSmsCode = function () {

        // 判断手机号码的有效性
        if ($scope.user.phone && /^1[3|4|5|7|8|9]\d{9}$/.test($scope.user.phone)) {
            // 发送异步请求
            baseService.sendGet("/user/sendSmsCode?phone="
                + $scope.user.phone).then(function (response) {
                // 获取响应数据
                if (response.data) {
                    // 倒计时 (扩展)
                    $scope.flag = true;
                    // 调用倒计时方法
                    $scope.downcount(90);
                } else {
                    alert("获取短信验证码失败！");
                }
            });
        } else {
            alert("手机号码不正确！");
        }
    };


    // 倒计时方法
    $scope.downcount = function (seconds) {
        if (seconds > 0) {
            seconds--;
            $scope.tipMsg = seconds + "秒，后重新获取";

            // 开启定时器
            $timeout(function () {
                $scope.downcount(seconds);
            }, 1000);
        } else {
            $scope.tipMsg = "获取短信验证码";
            $scope.flag = false;
        }
    };

});